src/frontend/src/scheduler/worker_node_manager.rs

```diff
@@ -140,63 +145,217 @@ impl WorkerNodeManager {
+    ) -> SchedulerResult<ParallelUnitMapping> {
+        if let Some(pu_mapping) = self
+            .inner
+            .read()
+            .unwrap()
+            .get_serving_fragment_mapping(fragment_id)
+        {
+            return Ok(pu_mapping);
+        }
+        let mut guard = self.inner.write().unwrap();
+        if let Some(pu_mapping) = guard.get_serving_fragment_mapping(fragment_id) {
+            return Ok(pu_mapping);
+        }
+        let pu_mapping = guard.reschedule_serving(fragment_id)?;
+        Ok(pu_mapping)
+    }
+}
+
+impl WorkerNodeManagerInner {
+    fn get_serving_fragment_mapping(&self, fragment_id: FragmentId) -> Option<ParallelUnitMapping> {
+        self.serving_fragment_vnode_mapping
+            .get(&fragment_id)
+            .cloned()
+    }
+
+    /// Calculates serving vnode mappings for `fragment_id`.
+    /// The `serving_parallelism` is set to min(`all_serving_pus`, `streaming_parallelism`).
+    fn reschedule_serving(
+        &mut self,
+        fragment_id: FragmentId,
+    ) -> SchedulerResult<ParallelUnitMapping> {
+        let all_serving_pus: BTreeMap<u32, Vec<ParallelUnit>> = self
+            .worker_nodes
+            .iter()
+            .filter(|w| w.property.as_ref().map_or(false, |p| p.is_serving))
+            .map(|w| (w.id, w.parallel_units.clone()))
+            .collect();
+        let serving_pus_total_num = all_serving_pus.values().map(|p| p.len()).sum::<usize>();
+        if serving_pus_total_num == 0 {
+            return Err(SchedulerError::EmptyWorkerNodes);
+        }
+        let streaming_vnode_mapping = self
+            .streaming_fragment_vnode_mapping
+            .get(&fragment_id)
+            .ok_or_else(|| {
+                anyhow!(
+                    "streaming vnode mapping for fragment {} not found",
+                    fragment_id
+                )
+            })?;
+        let serving_parallelism = std::cmp::min(
+            serving_pus_total_num,
+            streaming_vnode_mapping.iter_unique().count(),
+        );
+        assert!(serving_parallelism > 0);
+        // round-robin similar to `Scheduler` in meta node.
+        let mut parallel_units: LinkedList<_> = all_serving_pus
+            .into_values()
+            .map(|v| v.into_iter().sorted_by_key(|p| p.id))
+            .collect();
+        let mut round_robin = Vec::new();
+        while !parallel_units.is_empty() {
+            parallel_units.drain_filter(|ps| {
+                if let Some(p) = ps.next() {
+                    round_robin.push(p);
+                    false
+                } else {
+                    true
+                }
+            });
+        }
+        round_robin.truncate(serving_parallelism);
+        round_robin.sort_unstable_by_key(|p| p.id);
+        let selected_serving_pus = ParallelUnitMapping::build(&round_robin);
+        self.serving_fragment_vnode_mapping
+            .insert(fragment_id, selected_serving_pus.clone());
```

Comment:
So basically we are randomly picking a node to serve a request? I think in the original design we will push storage to serving node?

Reply:
This way how do we guarantee the latency of serving?

Reply:
> So basically we are randomly picking a node to serve a request?
This method is called once, and then we'll cache the node mapping for future use to achieve locality.

Reply:
> in the original design we will push storage to serving node

We cannot offer independent availability for streaming and serving this way.

Reply:
But will this code break current locality guarantee?